pkg/ecdsa/dkg/protocol.go

```diff
@@ -115,5 +117,318 @@ func (skgm *symmetricKeyGeneratingMember) isValidEphemeralPublicKeyMessage(
 	return true
 }
 
-// TODO: Implement other phases of the protocol that involve actual
-//       key generation.
+// tssRoundOne starts the TSS process by executing its first round. The
+// outcome of that round is a message containing commitments and Paillier
+// public keys for all other group members.
+func (trom *tssRoundOneMember) tssRoundOne(
+	ctx context.Context,
+) (*tssRoundOneMessage, error) {
+	if err := trom.tssParty.Start(); err != nil {
+		return nil, fmt.Errorf(
+			"failed to start TSS round one: [%v]",
+			err,
+		)
+	}
+
+	// We expect exactly one TSS message to be produced in this phase.
+	select {
+	case tssMessage := <-trom.tssOutgoingMessagesChan:
+		tssMessageBytes, _, err := tssMessage.WireBytes()
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to encode TSS round one message: [%v]",
+				err,
+			)
+		}
+
+		return &tssRoundOneMessage{
+			senderID:  trom.id,
+			payload:   tssMessageBytes,
+			sessionID: trom.sessionID,
+		}, nil
+	case <-ctx.Done():
+		return nil, fmt.Errorf(
+			"TSS round one outgoing message was not generated on time",
+		)
+	}
+}
+
+// tssRoundTwo performs the second round of the TSS process. The outcome of
+// that round is a message containing shares and de-commitments for all other
+// group members.
+func (trtm *tssRoundTwoMember) tssRoundTwo(
+	ctx context.Context,
+	tssRoundOneMessages []*tssRoundOneMessage,
+) (*tssRoundTwoMessage, error) {
+	// Use messages from round one to update the local party and advance
+	// to round two.
+	for _, tssRoundOneMessage := range tssRoundOneMessages {
+		senderID := tssRoundOneMessage.SenderID()
+		sortedPartiesIDs := trtm.tssParameters.Parties().IDs()
+		senderPartyIDKey := memberIDToTssPartyIDKey(senderID)
+		senderPartyID := sortedPartiesIDs.FindByKey(senderPartyIDKey)
+
+		_, tssErr := trtm.tssParty.UpdateFromBytes(
```

Comment:
When do we use named errors rather than `err` in the code base? There must be times when we need to keep around two errors simultaneously rather than handling one and overriding the variable.

Here though, it looks like we could call this just `err` since there aren't any other errors in scope

Reply:
This is because `UpdateFromBytes` doesn't return a pure `error` type but `tss.Error` instead. In several places, I needed to differentiate between them. I decided to do the same in all cases for consistency.